tizen-api/src/tizen-api-pipeline.c

```diff
@@ -0,0 +1,572 @@
+
+  num_mems = gst_buffer_n_memory (b);
+
+  if (num_mems > NNS_TENSOR_SIZE_LIMIT) {
+    dlog_print (DLOG_ERROR, "nnstreamer-capi-pipeline",
+        "Number of memory chunks in a GstBuffer exceed the limit: %u > %u",
+        num_mems, NNS_TENSOR_SIZE_LIMIT);
+    return;
+  }
+
+  for (i = 0; i < num_mems; i++) {
+    mem[i] = gst_buffer_peek_memory (b, i);
+    gst_memory_map (mem[i], &info[i], GST_MAP_READ);
+    buf[i] = (const char *) info[i].data;
+    size[i] = info[i].size;
+    total_size += size[i];
+  }
+
+  g_mutex_lock (&elem->lock);
+
+  /** @todo This assumes that padcap is static */
+  if (elem->sink == NULL) {
+    /* Get the sink-pad-cap */
+    elem->sink = gst_element_get_static_pad (elem->element, "sink");
+
+    if (elem->sink) {
+      /* sinkpadcap available (negotiated) */
+      GstCaps *caps = gst_pad_get_current_caps (elem->sink);
+
+      if (caps) {
+        guint n_caps = gst_caps_get_size (caps);
+        GstTensorsConfig tconfig;
+        gboolean found = FALSE;
+
+        for (i = 0; i < n_caps; i++) {
+          GstStructure *s = gst_caps_get_structure (caps, i);
+
+          found = gst_tensors_config_from_structure (&tconfig, s);
+          if (found)
+            break;
+        }
+
+        if (found) {
+          memcpy (&elem->tensorsinfo, &tconfig.info, sizeof (GstTensorsInfo));
+          elem->size = 0;
+
+          g_assert (elem->tensorsinfo.num_tensors == num_mems);
+
+          for (i = 0; i < elem->tensorsinfo.num_tensors; i++) {
+            size_t sz = gst_tensor_info_get_size (&elem->tensorsinfo.info[i]);
+
+            g_assert (sz == size[i]);
+            elem->size += sz;
+          }
+        } else {
+          elem->sink = NULL;    /* It is not valid */
+          /** @todo What if it keeps being "NULL"? Exception handling at 2nd frame? */
+        }
+      }
+    }
+  }
+
+  g_assert (gst_buffer_get_size (b) == total_size);
+  if (elem->size > 0)
+    g_assert (gst_buffer_get_size (b) == elem->size);
+
+  /* Iterate e->handles, pass the data to them */
+  for (l = elem->handles; l != NULL; l = l->next) {
+    nns_sink *sink = l->data;
+    nns_sink_cb callback = sink->cb;
+
+    callback (buf, size, &elem->tensorsinfo, sink->pdata);
+
+    /** @todo Measure time. Warn if it takes long. Kill if it takes too long. */
+  }
+
+  g_mutex_unlock (&elem->lock);
+
+  for (i = 0; i < num_mems; i++) {
+    gst_memory_unmap (mem[i], &info[i]);
+  }
+
+  return;
+}
+
+/**
+ * @brief Private function for nns_pipeline_destroy, cleaning up nodes in namednodes
+ */
+static void
+cleanup_node (gpointer data)
+{
+  element *e = data;
+  g_mutex_lock (&e->lock);
+  g_free (e->name);
+  if (e->src)
+    gst_object_unref (e->src);
+  if (e->sink)
+    gst_object_unref (e->sink);
+
+  /** @todo CRITICAL. Stop the handle callbacks if they are running/ready */
+  if (e->handles)
+    g_list_free (e->handles);
+  e->handles = NULL;
+
+  g_mutex_unlock (&e->lock);
+  g_mutex_clear (&e->lock);
+
+  g_free (e);
```

Comment:
free `e->tensorsinfo.info[*].name` ?

Reply:
```e->tensorsinfo.info[*].name``` can only be created by 
```memcpy (&elem->tensorsinfo, &tconfig.info, sizeof (GstTensorsInfo)); ``` in Line 116
, which simply copies the address from tconfig, which does NOT set name.

Therefore, name in this context is always NULL. :)

Reply:
Yes.